ext/hash/php_hash_sha.h

```diff
@@ -56,10 +56,10 @@ PHP_HASH_API void PHP_SHA256Update(PHP_SHA256_CTX *, const unsigned char *, size
 void SHA256_Transform_sse2(uint32_t state[PHP_STATIC_RESTRICT 8], const uint8_t block[PHP_STATIC_RESTRICT 64], uint32_t W[PHP_STATIC_RESTRICT 64], uint32_t S[PHP_STATIC_RESTRICT 8]);
 #endif
 
-#if (defined(__i386__) || defined(__x86_64__)) && defined(HAVE_IMMINTRIN_H)
+#if ((defined(__i386__) || defined(__x86_64__)) && defined(HAVE_IMMINTRIN_H)) || ((defined(_M_X64) || defined(_M_IX86)) && defined(ZEND_INTRIN_SSSE3_RESOLVER))
 # if defined(__SSSE3__) && defined(__SHA__)
 #  define PHP_HASH_INTRIN_SHA_NATIVE 1
-# elif defined(HAVE_FUNC_ATTRIBUTE_TARGET)
+# elif defined(HAVE_FUNC_ATTRIBUTE_TARGET) || defined(_M_X64) || defined(_M_IX86)
```

Comment:
I wonder if this should be an `# else` instead. Currently the code defines the prototype for `SHA256_Transform_shani()` even if neither the NATIVE, nor the RESOLVER functionality is used.

Reply:
Not sure if I understand you. I thought that `_M_X64` and `_M_IX86` are MSVC specific, and the prototype is not compiled unless `defined(HAVE_FUNC_ATTRIBUTE_TARGET)`, which is never the case with MSVC.

Reply:
Let me try to rephrase:

It might be possible (on Linux) that neither PHP_HASH_INTRIN_SHA_NATIVE nor PHP_HASH_INTRIN_SHA_RESOLVER are defined (this is a pre-existing issue). Namely when the compiler does neither support the `target()` attribute, nor has `__SHA__` defined.

The `SHA256_Transform_shani` would still be defined, but the implementation would then be missing, because the PR changes it to be guarded by `PHP_HASH_INTRIN_SHA_NATIVE || PHP_HASH_INTRIN_SHA_RESOLVER`.

It might make sense to move the prototype out of the existing `#if` into a new `#if defined(PHP_HASH_INTRIN_SHA_NATIVE) || defined(PHP_HASH_INTRIN_SHA_RESOLVER)` for consistency with the condition in `hash_sha_ni.c`.

Reply:
Now merged, you'll likely need to rebase this PR.